### ISSUE TITLE
[YSQL] YB Batched Nested Loop Join errors with a collation error when using an index

### ISSUE THREAD

**Issue by jameshartig** [yugabyte/yugabyte-db#15847]
Jira Link: [DB-5243](https://yugabyte.atlassian.net/browse/DB-5243)
### Description

When joining two tables on a TEXT column, an index is involved, and setting `yb_bnl_batch_size` > 1 then the query returns an error `could not determine which collation to use for string comparison`. Tested with `2.16.0.0-b90` and `2.16.0.1-b7`.

Setup a database and switch to it
```sql
CREATE DATABASE test;
\c test;
```

Create some tables
```sql
CREATE TABLE IF NOT EXISTS a (
	id TEXT NOT NULL,
	name TEXT NOT NULL,
	PRIMARY KEY(id)
);
CREATE TABLE IF NOT EXISTS b (
	a_id TEXT NOT NULL,
	user_id TEXT,
	PRIMARY KEY(a_id, user_id)
);
```

Insert some fake data
```sql
INSERT INTO a (id, name) VALUES ('1', 'foo');
INSERT INTO b (a_id, user_id) VALUES ('1', '2');
INSERT INTO b (a_id, user_id) VALUES ('2', '2');
```

Observe that everything is fine without an index:
```sql
EXPLAIN SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
/*+ SET(yb_bnl_batch_size 5) NestLoop(a b) */ EXPLAIN SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
/*+ SET(yb_bnl_batch_size 5) NestLoop(a b) */ SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
```

The query plan without `yb_bnl_batch_size` is:
```
                              QUERY PLAN
-----------------------------------------------------------------------
 Nested Loop  (cost=0.00..216.38 rows=1000 width=32)
   ->  Seq Scan on a  (cost=0.00..100.00 rows=1000 width=32)
   ->  Index Scan using b_pkey on b  (cost=0.00..0.12 rows=1 width=32)
         Index Cond: ((a_id = a.id) AND (user_id = '2'::text))
(4 rows)
```

The query plan with `yb_bnl_batch_size` is:
```
                                         QUERY PLAN
--------------------------------------------------------------------------------------------
 YB Batched Nested Loop Join  (cost=0.00..228.87 rows=1000 width=32)
   Join Filter: (a.id = b.a_id)
   ->  Seq Scan on a  (cost=0.00..100.00 rows=1000 width=32)
   ->  Index Scan using b_pkey on b  (cost=0.00..0.12 rows=1 width=32)
         Index Cond: ((a_id = ANY (ARRAY[a.id, $1, $2, $3, $4])) AND (user_id = '2'::text))
(5 rows)
```
but I had to force NestLoop otherwise it does a hash join. I'm not sure if this is fixed by https://github.com/yugabyte/yugabyte-db/issues/14240 or not.

Now add an index:
```sql
CREATE INDEX IF NOT EXISTS b_user_id ON b(user_id ASC) INCLUDE (a_id);
```

Observe new query plans:
```sql
EXPLAIN SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
/*+ SET(yb_bnl_batch_size 5) */ EXPLAIN SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
```

Without `yb_bnl_batch_size`:
```sql
                                   QUERY PLAN
--------------------------------------------------------------------------------
 Nested Loop  (cost=0.00..6.27 rows=10 width=32)
   ->  Index Only Scan using b_user_id on b  (cost=0.00..5.12 rows=10 width=32)
         Index Cond: (user_id = '2'::text)
   ->  Index Scan using a_pkey on a  (cost=0.00..0.11 rows=1 width=32)
         Index Cond: (id = b.a_id)
(5 rows)
```

With `yb_bnl_batch_size`:
```sql
                                   QUERY PLAN
--------------------------------------------------------------------------------
 YB Batched Nested Loop Join  (cost=0.00..6.39 rows=10 width=32)
   Join Filter: (a.id = b.a_id)
   ->  Index Only Scan using b_user_id on b  (cost=0.00..5.12 rows=10 width=32)
         Index Cond: (user_id = '2'::text)
   ->  Index Scan using a_pkey on a  (cost=0.00..0.11 rows=1 width=32)
         Index Cond: (id = ANY (ARRAY[b.a_id, $1, $2, $3, $4]))
(6 rows)
```
and I don't need the NestJoin hint anymore.

Verify the non-batch query works:
```sql
SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
```

but now the batch query fails with a collation error:
```
# /*+ SET(yb_bnl_batch_size 5) */ SELECT a.id FROM a LEFT JOIN b ON a.id = b.a_id WHERE b.user_id = '2';
ERROR:  could not determine which collation to use for string comparison
HINT:  Use the COLLATE clause to set the collation explicitly.
```

[DB-5243]: https://yugabyte.atlassian.net/browse/DB-5243?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by FranckPachot** [comment]
@jameshartig  according to @tanujnay112, this should be fixed in next version 2.17.2
Also the need to hint should be solved as the cost optimizer will handle it. List of improvements: https://github.com/yugabyte/yugabyte-db/issues/14070

**Comment by tanujnay112** [comment]
Hi @jameshartig
Sorry about the delayed response! This feature is considered beta until 2.18.
